Gallery block: turn on auto-migration of v1 Gallery blocks to v2 format when edited

glendaviesnz opened the pull request:
## Description

Turns on automigration of v1 gallery blocks to v2 format unless a site has `use_BalanceTags`  enabled and is not running WP >= 5.9

## Testing
- **Before checking out this PR**, on a site on WP < 5.9 set use_balanceTags option to 1 
- Then check out this PR and add a gallery and check that it is still the old format (see screenshots below) when editing and viewing and when reloading in the editor
- Switch use_balanceTags to 0 and  load the v1 Gallery block in the editor and check that it auto migrates to the v2 block format (see screenshots below) and can be edited, saved, reloaded ok
- Try setting `use_balanceTags` back to 1 and you should get an error message about needing to upgrade to WP 5.9
- For both v1 and v2 blocks check that you can transform to individual Image blocks and back to Gallery block again, and when transforming back to Gallery make sure it is v1 format for `use_balanceTags` site, and v2 for others 
- Also try adding a v1 block with external images (to do this add Image blocks and use the replace button to add the url of an external image - then transform the Image blocks to a Gallery block) and make sure these migrate to v2 gallery format ok. There is a [known Image size selector error](https://github.com/WordPress/gutenberg/issues/36554) with this flow

**Important note**: It is not possible to load v2 version blocks in a site with `use_balanceTags` on and WP < 5.9, they will show an invalid error. The plan is to add some sort of explanatory error to indicate that user has to turn off `use_balanceTags` or upgrade to 5.9 in these instances - there is no point in showing a valid block as it will just break when the user saves it, but the user will not know unless they reload or view in the frontend, so better to fail on first editor load.

You can use `wp option set use_balanceTags 1` to toggle the option, or you can make the option available at `/wp-admin/options-writing.php` by running `wp option set initial_db_version 32452`

## Screenshots

v1 Gallery block:
<img width="664" alt="Screen Shot 2021-11-16 at 9 50 57 PM" src="https://user-images.githubusercontent.com/3629020/142125678-724eabc5-c9a3-4b9a-96ea-289278ccb2b3.png">

v2 Gallery block:
<img width="645" alt="Screen Shot 2021-11-16 at 9 58 45 PM" src="https://user-images.githubusercontent.com/3629020/142125871-0645fea7-7ae6-4ce0-a715-eab6f9ee8c60.png">

Commit:
Update experimental flag to block sites with use_BalanceTags option enabled and aren't on WP 5.9 or higher

Commit:
Add check for gallery v2 compat to editor init so it is available when block deprecation pipeline runs

Commit:
Remove references to gallery experimental flag

Commit:
Revert "Remove references to gallery experimental flag"

This reverts commit b988d24cde235e131224123b98b8d93b1ffd8940.

Commit:
Abstract out the check for v2 gallery support and use the existing experimental flag for mobile

Commit:
Add missing transformation updates

Commit:
Update fixture to be compat with v2 of gallery block

Commit:
Fix unit tests

Commit:
Fix issue with link destination not being set when block migrated

mkevins commented on packages/block-library/src/gallery/shared.js:
```diff
@@ -21,3 +28,20 @@ export const pickRelevantMediaFiles = ( image, sizeSlug = 'large' ) => {
+	}
+
+	const settings = select( blockEditorStore ).getSettings();
+	if ( settings.__unstableGalleryWithImageBlocks ) {
+		return true;
+	}
+
+	return false;
```
NP: at this point, we can just:

```
return !!settings.__unstableGalleryWithImageBlocks
```

i.e. without the conditional.

glendaviesnz replied:
Doh! - yeh that makes more sense - fixed.

mkevins commented:
Hi Glen :wave: 

I encountered an issue when testing this on mobile. Steps:

1. Created a JN site
2. Set use_balanceTags on via: `wp option set use_balanceTags 1`
3. Added this site to my mobile app
4. Created a new post with a gallery (confirmed that it was using the original format :heavy_check_mark: )
5. Saved and exited the post
6. Previewed the post (confirmed that the images were still present and the gallery is intact :heavy_check_mark: )
7. Reopened the post (:x: gallery shows "problem displaying block")
8. Checked HTML view (:x: confirmed that the images are missing)

Gallery Visual | Gallery HTML
|-|-|
![gallery-created-visual](https://user-images.githubusercontent.com/8507675/141256421-9ee2224e-a2eb-42c8-85bb-fb1602bc680b.png)|![gallery-created-html](https://user-images.githubusercontent.com/8507675/141256403-75325dcf-1c30-4ab4-a678-8feb7a02b534.png)

**Preview of Gallery:**
<img src="https://user-images.githubusercontent.com/8507675/141256429-a811d6a0-4f30-48cc-baf9-c98c90f42a48.png" width="360">

Re-opened Gallery Visual | Re-opened Gallery HTML
-|-
![gallery-reopened-visual](https://user-images.githubusercontent.com/8507675/141256440-fd2b1851-e9b8-4234-b170-2c290dbbd749.png)|![gallery-reopened-html](https://user-images.githubusercontent.com/8507675/141256436-5ceb7515-eec1-43a4-a3d4-6ccf36fac145.png)

So, seemingly something is destroying the gallery on when the editor is opened the second time (maybe in the deprecations flow :thinking: ? ).

Commit:
Check that we have a v1 gallery block before we send content through the v1 save function

mkevins committed:
Use `window.wp` global instead of store for gallery flag

To resolve a race with the deprecations and asynchronous store updates
for the gallery flag, we can use a property on the global wp object
instead.

mkevins committed:
Remove unnecessary destructuring

This isn't necessary, since `getThemeColros` already returns a new
object.

Commit:
Move the setting of wp.galleryBlockV2Enabled to directly after wp-dom-ready to be sure it is set well ahead of the block deprecations running

Commit:
Add optional chaining operator to prevent error if images attribute not defined

Commit:
Simplify deprecations, etc. by always using using v1 methods if use_balanceTags is on regardless of content format

Commit:
Fix potential issue with attributes not being returned from v6 deprecation

andrewserong commented on packages/block-library/src/gallery/shared.js:
```diff
@@ -21,3 +21,11 @@ export const pickRelevantMediaFiles = ( image, sizeSlug = 'large' ) => {
 };
+
+export function isGalleryV2Enabled() {
+	if ( typeof process !== 'undefined' && process.env?.NODE_ENV === 'test' ) {
```
This is possibly a bit nit-picky, and I imagine it doesn't matter too much since it's a simple check. But I was wondering if this needs to be in run-time code. Would it be possible to set `window.wp.galleryBlockV2Enabled` to `true` via globals in Jest config instead? I don't think it's a blocker, though, because the check is neatly tucked away inside the `isGalleryV2Enabled` function so would be easy to change later if we want to.

glendaviesnz replied:
🤦 yeh, jest globals is a much better option - completely overlooked that for some reason when trying to set this for test runs! Have moved there.

andrewserong commented on packages/editor/src/components/provider/index.native.js:
```diff
@@ -97,9 +101,11 @@ class NativeEditorProvider extends Component {
 		} = this.props;
 
+		// TODO: remove this as unnecessary since we are setting in constructor?
+		window.wp.galleryBlockV2Enabled = galleryWithImageBlocks;
```
I haven't tested this, but it looks like it can be removed if it's set in the constructor?

mkevins replied:
Thanks Andrew! That was a WIP "note to self" :smiley:  that I intended to circle back to and make sure I wan't missing anything about when this is set from the fetch response (originally this was setting a value in the store via `updateSettings`, which couldn't be done in the constructor - probably the reason it was set after mount initially). :+1:

andrewserong commented on test/integration/__snapshots__/blocks-raw-handling.test.js.snap:
```diff
@@ -57,8 +57,8 @@ exports[`rawHandler should convert HTML post to blocks with minimal content chan
-<!-- wp:gallery {\\"ids\\":[1],\\"columns\\":3,\\"linkTo\\":\\"post\\"} -->
-<figure class=\\"wp-block-gallery columns-3 is-cropped\\"><ul class=\\"blocks-gallery-grid\\"><li class=\\"blocks-gallery-item\\"><figure><img data-id=\\"1\\" class=\\"wp-image-1\\"/></figure></li></ul></figure>
+<!-- wp:gallery {\\"shortCodeTransforms\\":[{\\"id\\":1}],\\"columns\\":3,\\"linkTo\\":\\"none\\"} -->
+<figure class=\\"wp-block-gallery has-nested-images columns-3 is-cropped\\"></figure>
```
This diff looks like it now strips the internal `img` element. Is this an error, similar to the migration issue I ran into? I imagine we should be seeing an image block within the figure element, here.

glendaviesnz replied:
For the shortcode transforms with the v2 gallery block the transform has to just add the ids to a temporary `shortCodeTransforms` attribute as the transform method doesn't have access to all the data it needs to convert these into innerBlocks - the innerBlocks `<img>` elements are created by a `useEffect` in  the Gallery block component once it mounts - but these tests are not mounting the actual block code, they are just running `pasteHandler` code, so the `img` elements are not generated.

andrewserong replied:
Ah, that makes sense, now. Thanks for explanation, Glen!

andrewserong commented on test/integration/fixtures/documents/shortcode-matching-out.html:
```diff
@@ -1,7 +1,7 @@
-<!-- wp:gallery {"ids":[1000],"columns":3,"linkTo":"post"} -->
-<figure class="wp-block-gallery columns-3 is-cropped"><ul class="blocks-gallery-grid"><li class="blocks-gallery-item"><figure><img data-id="1000" class="wp-image-1000"/></figure></li></ul></figure>
+<!-- wp:gallery {"shortCodeTransforms":[{"id":1000}],"columns":3,"linkTo":"none"} -->
+<figure class="wp-block-gallery has-nested-images columns-3 is-cropped"></figure>
```
This appears to be missing an internal `img` element or block?

glendaviesnz replied:
same as above

andrewserong commented on test/integration/fixtures/documents/wordpress-out.html:
```diff
@@ -18,8 +18,8 @@ <h3>More tag</h3>
-<!-- wp:gallery {"ids":[1],"columns":3,"linkTo":"post"} -->
-<figure class="wp-block-gallery columns-3 is-cropped"><ul class="blocks-gallery-grid"><li class="blocks-gallery-item"><figure><img data-id="1" class="wp-image-1"/></figure></li></ul></figure>
+<!-- wp:gallery {"shortCodeTransforms":[{"id":1}],"columns":3,"linkTo":"none"} -->
+<figure class="wp-block-gallery has-nested-images columns-3 is-cropped"></figure>
```
This also appears to be missing an internal `img` element or block?

glendaviesnz replied:
same as above

andrewserong commented on packages/block-library/src/gallery/shared.js:
```diff
@@ -21,3 +21,11 @@ export const pickRelevantMediaFiles = ( image, sizeSlug = 'large' ) => {
+		return true;
+	}
+
+	return window.wp.galleryBlockV2Enabled;
```
Tiniest of tiny nits, but since this function is called `isGalleryV2Enabled`, should this return a boolean (e.g. `!!`)? Also, I assume that the `wp` object will always be available in practice, but I see that elsewhere in the Gallery block where `window.wp` is accessed, we've used optional chaining. Not sure if it's needed here, too?

mkevins replied:
> should this return a boolean

Hm.. `window.wp.galleryBlockV2Enabled` already is a boolean, I wonder, do you mean to coerce with `!!` for the case where it is `undefined`?

> Not sure if it's needed here, too?

Good question.. I had contemplated that as well.. my thinking at the time (at least from the mobile perspective) is that `window.wp` should be defined (and as an `Object`), and that since we are altering user content on the basis of this flag, it'd be better to fail fast if that assumption is incorrect, to avoid incorrectly transforming the content. Otoh, this is function is shared cross-platform, so it'd be good to know from the web side whether this assumption is valid - and whether or not failing fast is appropriate there as well in case it's `undefined`. :thinking: 

Also, maybe it's actually fine on mobile and web to default to false(y) when the assumption is not met? :thinking: One of the hesitations I originally had was born out of the race conditions we've encountered while testing / debugging these flows - not sure whether the presence (or absence) of `window.wp`, if even possible, could also be a timing-based condition?

glendaviesnz replied:
@mkevins, @andrewserong I wonder if we want to default to v2 being enabled if `window.wp.galleryBlockV2Enabled` is not explicitly set to `false`, eg. `return window.wp?.galleryBlockV2Enabled === false ? false : true;`  - the reason being that the disabled case with `use_balanceTags === 1` and `WP < 5.9` is the edge case so are we better to assume we use v2 if `wp.galleryBlockV2Enabled` was undefined for some reason?

andrewserong replied:
Thanks for the thoughts Matthew and Glen!

>  I wonder, do you mean to coerce with `!!` for the case where it is `undefined`?

Yes, I was mostly thinking of how do we want to handle the case of `undefined`.

> I wonder if we want to default to v2 being enabled if `window.wp.galleryBlockV2Enabled` is not explicitly set to false

That logic sounds good to me, as you say, the presence of the flag is there to deal with an explicit edge case, so handling it with an explicit check for `false` sounds good to me 👍

andrewserong reviewed (commented):
Nice work @glendaviesnz! This is working well for migrating from current v1 markup to v2, while preserving columns settings and "link to" settings 👍

I also tested each of the `core__gallery__deprecated-x.html` deprecations in the v1 version of the editor. Initially I thought I ran into an issue, but then realised I hadn't pulled in the latest changes, but as of https://github.com/WordPress/gutenberg/pull/36191/commits/e43a6a720c9fadc2ec9954560154338c445c34a8 each of the deprecations is migrating correctly when switching between the code and visual editor views 👍 

It's all working well for me in practice. Looking at the diff, the main issue I noticed is that a couple of the fixtures (`shortcode-matching-out.html` and `wordpress-out.html`) and a snapshot appear to be stripping the internal `img` element in the diff, but I wasn't quite sure how best to test them in practice, so not sure if it's an issue? Other than that, I left a couple of tiny nits, but feel free to ignore if they're not relevant!

mkevins committed:
Move assignment of gallery global flag to native Editor

This moves the setting of the global flag higher in the view hierarchy,
since the provider's constructor is still too late (parsing code invokes
the deprecations before even the provider is instantiated).

Commit:
Move mocking of the galleryBlockV2Enabled flag to the jest globals file

Commit:
Explicitly check for galleryBlockV2Enabled being false and default to true in other instances

Commit:
Fail early if window.wp.galleryBlockV2Enabled  is not a boolean

andrewserong commented on packages/block-library/src/gallery/shared.js:
```diff
@@ -21,3 +21,19 @@ export const pickRelevantMediaFiles = ( image, sizeSlug = 'large' ) => {
+export function isGalleryV2Enabled() {
+	// We want to fail early here, at least during beta testing phase, to ensure
+	// there aren't instances where undefined values cause false negatives.
+	if ( ! window.wp || typeof window.wp.galleryBlockV2Enabled !== 'boolean' ) {
```
Just wondering, if this PR is backported to the WP core beta, then will the logic in `gutenberg_check_gallery_block_v2_compatibility` also be backported? Without it, in 5.9, with Gutenberg deactivated, where `wp.galleryBlockV2Enabled` isn't set, will this throw an error?

In that case, it could be better to treat the v2 gallery block setting being `undefined` as `true` 🤔. I might be missing some context with the impact on mobile, though!

mkevins replied:
Good question, Andrew. For the mobile perspective, I'm going to test making `true` default (because some users will not have a cached value when the editor is initialized):

```diff
diff --git a/packages/edit-post/src/editor.native.js b/packages/edit-post/src/editor.native.js
index b045724a04..7f71d7f68a 100644
--- a/packages/edit-post/src/editor.native.js
+++ b/packages/edit-post/src/editor.native.js
@@ -31,7 +31,9 @@ class Editor extends Component {
 		super( ...arguments );
 
 		// need to set this globally to avoid race with deprecations
-		window.wp.galleryBlockV2Enabled = props.galleryWithImageBlocks;
+		// defaulting to true to avoid issues with a not-yet-cached value
+		const { galleryWithImageBlocks = true } = props;
+		window.wp.galleryBlockV2Enabled = galleryWithImageBlocks;
 
 		if ( props.initialHtmlModeEnabled && props.mode === 'visual' ) {
 			// enable html mode if the initial mode the parent wants it but we're not already in it
```

For mobile, we won't face the same backporting issue, since users on older versions won't encounter this logic, and users on newer versions will default to `true` (assuming all goes well with testing :point_up_2: ). I wonder if something similar would work (or be appropriate) for web? Or perhaps that is an error that we'd like to surface in beta (if it exists)?

andrewserong replied:
Thanks Matthew, nice to see that's how we'll handle the situation in mobile 👍. Let's see what @glendaviesnz thinks for the web when he's back on tomorrow 🙂

glendaviesnz replied:
@andrewserong `gutenberg_check_gallery_block_v2_compatibility` will also need to be pulled into WP 5.9. The explicit error we are throwing here is mostly for test purposes to see if we ever hit that explicit exception during testing - I wasn't thinking we would leave that in in there permanently, but instead more likely default to `true` once we are comfortable there aren't any funny edge cases we have overlooked that would have that value as `undefined` when it really should be `false` - does that make sense?

andrewserong replied:
Yes, thanks for confirming @glendaviesnz! That does sound like a good way to handle it for the beta and so that issues don't go unnoticed.

It sounds like in that case we might want to open up a PR in [wordpress-develop](https://github.com/wordpress/wordpress-develop) to add in the `check_gallery_block_v2_compatibility` function prior to this PR landing, since I _think_ the PHP code gets merged in separately from the JS being released in npm packages for 5.9?

glendaviesnz replied:
@andrewserong `check_gallery_block_v2_compatibility` is in the gutenberg repo and [added as part of this PR](https://github.com/WordPress/gutenberg/pull/36191/files#diff-83705774c15f7e3df5b54b2096fe476721e8b962ba319a8e7dc46d7b9b0b27b4R207), so was thinking that it gets included in 5.9 when this PR merge is cherry picked? I don't see any of the other methods in `lib/compat.php` anywhere in the `wordpress-develop repo, am I missing something?

andrewserong replied:
@glendaviesnz good question — I'm not 100% sure of the order in which things go into 5.9. I believe there are two steps: commits for 5.9 are cherry picked into the [wp/trunk](https://github.com/wordpress/gutenberg/tree/wp/trunk) branch and then new npm package releases are published (e.g. [in this recent commit](https://github.com/WordPress/gutenberg/commit/ca22e8852fcc91889510e417bdaa180ad60f8dac)). Then, the next step is that over in `wordpress-develop` there's a PR created that updates trunk to point to the new packages (e.g. https://github.com/WordPress/wordpress-develop/pull/1883).

It looks like that latter PR includes some PHP changes in addition to bumping the package numbers. I'm not sure if these changes needed to be made manually? So for the case of `check_gallery_block_v2_compatibility`, it'd be good to confirm where it'll be merged (or how best to do it) before we merge this PR.

CC: @noisysocks — just a question for the 5.9 release, this PR includes an additional function `check_gallery_block_v2_compatibility` that currently lives in `lib/compat.php` in Gutenberg, but will need to exist in core. Do we need to open up a separate PR in `wordpress-develop` for this function, or will it get rolled in when the PR is cherry picked into 5.9?

noisysocks replied:
When I prepare the Core patch that updates packages (e.g. https://github.com/WordPress/wordpress-develop/pull/1883) every Monday, I do a diff of what PHP has changed in the Gutenberg plugin (`git diff $old_sha wp/trunk **/*.php`)  and then manually make those changes in Core.

For big PHP changes, I appreciate if the PR author can open a ticket in Core Trac and make those changes themselves. That makes it less likely I have to stay up late on Monday 😀

For smaller changes such as this one, it's okay to leave it with me. You can help me (or any future release lead) out by leaving a note in the doc comment which describes how the change should look once added to Core.

mkevins committed:
Default to `true` for gallery flag when not yet cached

This makes sure the global value for the flag is established when the
editor is initialized, since some users may not have the value cached
locally yet. This is important to avoid incorrectly branching in the
parsing step, resulting in an incompatible block format.

andrewserong reviewed (commented):
@glendaviesnz this PR is testing well for me, thanks for addressing all the global mock for `galleryBlockV2Enabled`, and it sounds like we're good to go in regards to merging in the PHP change 👍

I haven't been able to fault the changes in this PR, but when I went to test a v1 block with external images, I noticed that I'm unable to click any of the buttons in the v1 gallery block to replace an image with an external image. It's these Upload, Media Library, and Insert from URL buttons in the placeholder that don't appear to be clickable for me (only in the v1 version):

![image](https://user-images.githubusercontent.com/14988353/143155430-0e9fe892-7ff0-4de7-a804-aa41e142f1a7.png)

Are you able to replicate that? It appears to be unrelated to this PR, but meant that I wasn't _too_ sure how to test that external images were migrated correctly. Other than that, this change LGTM!

andrewserong commented:
> I noticed that I'm unable to click any of the buttons in the v1 gallery block to replace an image with an external image. It's these Upload, Media Library, and Insert from URL buttons in the placeholder that don't appear to be clickable for me (only in the v1 version):

I think I've worked out a fix for this @glendaviesnz, so I'll open up a PR soon 🙂
Edit: fix in https://github.com/WordPress/gutenberg/pull/36804

mkevins committed:
Only update gallery flag if fetch result is explicitly boolean

This adds the flag to the destructuring of the fetch callback (so it is
not binding to the stale closure value), and also conditionally updates
the global only if the type is explicitly boolean. This avoids a
scenario in which updates to other editor settings which omit the
gallery flag property result in the global value being overwritten with
`undefined`.

mkevins committed:
Update editor settings response to the editor to include gallery flag

This adds logic to ensure we report the gallery flag when the editor
settings are fetched.

mkevins commented:
I've added a few more commits to fix some issues with the flag (default value, cached value, and fetch response): https://github.com/WordPress/gutenberg/pull/36191/commits/0885b7aaa66f41fdd369a9fbbc0df0fe9352ed5c, https://github.com/WordPress/gutenberg/pull/36191/commits/73f556bcd96a5d4a76eae5573fd79927a1b60bc0,  https://github.com/WordPress/gutenberg/pull/36191/commits/1b6df54e057713e9b69c02a77a0ce6e2fdcb9b03.

I've also performed a bunch of tests on a Jurassic Ninja site (WordPress 5.8.2) w/ the plugin built locally from this branch installed:

## Gallery Auto-conversion Tests

To test the auto-conversion implementation, I used a test site with ssh access, toggled the remote flag and the locally cached flag, and opened posts with existing V1 and V2 content. The table below shows the results of these tests:

Starting with | Flag cached to | Remote flag | Result | New blocks
-|-|-|-|-
V1 content|`undefined`|`false`|:x: Content lost|V1
V1 content|`undefined`|`true`|V2|V2
V1 content|`false`|`false`|V1|V1
V1 content|`false`|`true`|:x: Content lost|V2
V1 content|`true`|`false`|:x: Content lost|V1
V1 content|`true`|`true`|V2|V2
V2 content|`undefined`|`false`|:x: Content lost|V1
V2 content|`undefined`|`true`|V2|V2
V2 content|`false`|`false`|:x: Content lost|V1
V2 content|`false`|`true`|:x: Content lost|V2
V2 content|`true`|`false`|:x: Content lost|V1
V2 content|`true`|`true`|V2|V2

In order to toggle the remote flag, the command in the WordPress directory is: `wp option set use_balanceTags 0` for `true` and `wp option set use_balanceTags 1` for `false` (Note: this is "inverted" because the gallery flag is _disabled_ when the `use_balanceTags` flag is _enabled_).

In order to manipulate the local cache, the App Inspection tab can be used in Android Studio. Under the `wp-fluxc` database, the flag is stored in the `EditorTheme` table. Since this table is regenerated each time the editor loads, it is safe to delete all the records to clear the cache with: `DELETE FROM EditorTheme;`. Also, it may be convenient to manually set the cached value to `true` or `false` by executing either `UPDATE EditorTheme SET GALLERY_WITH_IMAGE_BLOCKS=1;` or `UPDATE EditorTheme SET GALLERY_WITH_IMAGE_BLOCKS=0;` queries, respectively.

These tests have only been conducted for Android, though similar behavior should be expected for iOS, excepting that when the local cache is not populated, a default value of `false` [appears to be passed to the editor](https://github.com/WordPress/gutenberg/blob/50b19e70654c79e5f5c68a6ee97d1f3464aef1f0/packages/react-native-bridge/ios/Gutenberg.swift#L202). We can condition this, if desired, to leave it undefined over the bridge, which would result instead in the JS default value `true`.

Although many cases in this table result in content loss, the proportion of rows in this table in no way reflects the proportion of users / scenarios where these issues will be encountered. Most of these scenarios are edge cases involving the user toggling the deprecated `use_balanceTags` setting, or having it on while using the V2 content. For completeness, I conducted each permutation listed above, because there is potential for a race between editor initialization and fetch response (i.e. flag usage while parsing vs. editing blocks). It would be good to get some assurance about the impact of these scenarios. E.g. if usage of `use_balanceTags` is very minimal, I expect these issues to be encountered very rarely.

antonis commented:
@mkevins Thank you for the mobile fixes and the detailed testing and instructions 🙇 
I conducted I few tests today and I confirm your results 👍 

> Most of these scenarios are edge cases involving the user toggling the deprecated `use_balanceTags` setting, or having it on while using the V2 content.

I would agree with that. The content loss scenarios are edge cases and most of our users should not be affected.

mkevins committed:
Leave gallery flag `undefined` when not cached

This allows the value to be set to default `true` on the JS side,
instead of defaulting to a value of `false` from iOS native code.

Commit:
Update the wording of the mobile warning

Commit:
Add filter to prevent use_balanceTags being enabled on WP < 5.9

Commit:
Remove update gallery modal

Commit:
Switch to using settings_error for use_balanceTags notice

ramonjd commented:
I've run through the test instructions and it works as described. 

✅ Before checking out this PR, on a site on WP < 5.9 set use_balanceTags option to 1
✅ Then check out this PR and add a gallery and check that it is still the old format (see screenshots below) when editing and viewing and when reloading in the editor
✅ Switch use_balanceTags to 0 and load the v1 Gallery block in the editor and check that it auto migrates to the v2 block format (see screenshots below) and can be edited, saved, reloaded ok
✅ Try setting use_balanceTags back to 1 and you should get an error message about needing to upgrade to WP 5.9
✅ For both v1 and v2 blocks check that you can transform to individual Image blocks and back to Gallery block again, and when transforming back to Gallery make sure it is v1 format for use_balanceTags site, and v2 for others
✅ Also try adding a v1 block with external images (to do this add Image blocks and use the replace button to add the url of an external image - then transform the Image blocks to a Gallery block) and make sure these migrate to v2 gallery format ok. There is a known Image size selector error with this flow


> Try setting use_balanceTags back to 1 and you should get an error message about needing to upgrade to WP 5.9

Works on both the settings dashboard and CLI.

<img width="1261" alt="Screen Shot 2021-11-29 at 3 56 17 pm" src="https://user-images.githubusercontent.com/6458278/143811186-c5f9e0bb-f035-4591-9c88-36ff8751d22b.png">

Although the CLI doesn't provide any upgrade feedback

<img width="522" alt="Screen Shot 2021-11-29 at 3 24 14 pm" src="https://user-images.githubusercontent.com/6458278/143811230-ec440b2f-2a97-44c0-8f1f-0c4791340cf3.png">

I ran out of time to test more thoroughly, but will pick it up in the morning.

Commit:
Add a CLI error message

Commit:
Fix linting error

ramonjd commented:
Things are working well still for me!

CLI error appears with the latest change 🌮 

<img width="788" alt="Screen Shot 2021-11-30 at 11 36 00 am" src="https://user-images.githubusercontent.com/6458278/143968038-374d824c-7fb6-4885-8361-f58c8d5d2496.png">

I noticed one peculiarity: When I have `use_balanceTags === 1` (v1) activated, when I transform an Image to a v1 Gallery, and then select **Link to: Attachment page** in the dropdown, the attachment link doesn't appear.

![Kapture 2021-11-30 at 12 10 29](https://user-images.githubusercontent.com/6458278/143967637-b9a3e80f-45f4-4257-9a57-58ea098b64be.gif)

I ran out of time to debug (again) sorry, but would we need to do something similar to what we did in https://github.com/WordPress/gutenberg/pull/36388?

ramonjd commented on lib/compat.php:
```diff
@@ -191,3 +191,54 @@ function gutenberg_safe_style_attrs( $attrs ) {
+
+	if ( 1 === (int) $new_value && version_compare( $wp_version, '5.9', '<' ) ) {
+		/* translators: %s: Minimum required version */
+		$message = sprintf( __( 'Gutenberg requires WordPress %s or later in order to enable the &#8220;Correct invalidly nested XHTML automatically&#8221; option. Please upgrade WordPress before enabling.', 'gutenberg' ), '5.9' );
```
Not a biggie, but would it possible to strip or replace the HTML entities for the CLI?

<img width="788" alt="Screen Shot 2021-11-30 at 11 36 00 am" src="https://user-images.githubusercontent.com/6458278/143968789-804c0c8a-ee75-4383-971e-064f360e86c7.png">

glendaviesnz replied:
The chance of somebody toggling on this option via CLI with gutenberg 12.1+ on WP < 5.9 is so slim as to not warrant the bother of adding a string replace method here in my view.

ramonjd replied:
Sounds fair enough to me 👍

glendaviesnz commented:
> I noticed one peculiarity: When I have `use_balanceTags === 1` (v1) activated, when I transform an Image to a v1 Gallery, and then select **Link to: Attachment page** in the dropdown, the attachment link doesn't appear.

Nice catch! It looks like this is an existing bug - I get the same thing on 10.9.0, which is before the new Gallery was merged. It will take a bit of work to fix this as it requires for additional calls to getMedia after the transform in order to get the images `link` data. I am going to suggest that the fix for this is for the user to instead upgrade to 5.9 and use the new Gallery block - sound like a plan @ramonjd ?

ramonjd commented:
> I get the same thing on 10.9.0, which is before the new Gallery was merged. It will take a bit of work to fix this as it requires for additional calls to getMedia after the transform in order to get the images link data. I am going to suggest that the fix for this is for the user to instead upgrade to 5.9 and use the new Gallery block - sound like a plan @ramonjd ?

Oh great! Thanks for checking that it's an existing issue. In that case, it makes sense that we don't address it here and rely on the new gallery upgrade. Sounds good.

andrewserong reviewed (approved):
Great work again @glendaviesnz and @mkevins! This is testing well for me, and I haven't been able to find any faults with it 😀

✅ The auto-migration works correctly, irrespective of the `use_balanceTags` option when I'm running WP 5.9
✅ In 5.8.x, with `use_balanceTags` set to `1`, editing the gallery remains in the `v1` version of the gallery block
✅ In 5.8.x, if `use_balanceTags` is set to `0`, I'm unable to set it to `1` and the error message is rendered in the UI and in the CLI, and the UI message looks good:

![image](https://user-images.githubusercontent.com/14988353/144150899-cd4dc400-f61b-4473-8229-33ea68ab6e61.png)

✅ In 5.8.x, the v1 gallery is being correctly auto-migrated to the v2 format, with settings and external images preserved correctly
✅ Updated text for the app version notice when inserting a gallery block reads well to me

![image](https://user-images.githubusercontent.com/14988353/144151036-a14b3bfc-12fc-496f-91bd-184844789c4e.png)

This LGTM, and I did a quick visual skim of the code changes again. I'm not very familiar with React Native, but no issues jumped out at me from looking over the changes 🙂

ramonjd reviewed (approved):
I went through the same testing steps as before in https://github.com/WordPress/gutenberg/pull/36191#issuecomment-981291827

I also added a bunch of v1 galleries to a page, some with external images. All migrated as expected.

The v2 gallery on its own works as expected after migration as well.

🥇 Nice work!
